db/dcrpg/queries.go

```diff
@@ -3001,19 +3023,37 @@ func retrieveCoinSupply(ctx context.Context, db *sql.DB, timeArr []dbtypes.TimeD
 		var value int64
 		var timestamp time.Time
 		if err = rows.Scan(&timestamp, &value); err != nil {
-			return timeArr, sumArr, err
+			return timeArr, sumArr, estimateArr, err
 		}
 
 		if value < 0 {
 			value = 0
 		}
 
 		sum += dcrutil.Amount(value).ToCoin()
-		timeArr = append(timeArr, dbtypes.NewTimeDef(timestamp))
 		sumArr = append(sumArr, sum)
-	}
+		timeArr = append(timeArr, dbtypes.NewTimeDef(timestamp))
 
-	return timeArr, sumArr, nil
+		// if preminedCoins and initial time wasn't set it.
+		if len(sumArr) == 2 && preminedCoins == 0 {
+			preminedCoins = sum
+			initialTime = timestamp
+		}
+
+		var timeDiff = float64(timestamp.Sub(initialTime))
+		var estimatedBlockHeight = int(timeDiff / targerBlockTime)
+		if estimatedBlockHeight <= len(timeArr)-1 {
+			estimatedBlockHeight = len(timeArr) - 1
+		}
+
+		subsidyInterval := math.Floor(float64(estimatedBlockHeight) / subsidyReduction)
+		if value > 0 {
```

Comment:
`value` from the query is the cumulative coin supply at given timestamp.  So, if `value == 0`, that means it is the genesis block since every other block will add to the supply.  If that is correct, I suggest commenting on this logic here.

Also if `value == 0` on genesis, and the `estimate` is not updated, then `preminedCoins` should still be 0. I'm just talking through the logic here because I feel like this loop can be tweaked to make the code more expressive of the intent.